packages/affiliates/libs/bigquery.js

```diff
@@ -0,0 +1,82 @@
+const moment = require("moment");
+const assert = require("assert");
+const lodash = require("lodash");
+
+// Big query class meant for interacting with the ethereum table.
+module.exports = ({ client } = {}) => {
+  assert(client, "requires bigquery client");
+
+  // types of queries we are interested in
+  const queries = {
+    // logs aka events by contract
+    logsByContract(contract, start, end = Date.now(), selection = ["*"]) {
+      assert(contract, "requires contract");
+      assert(start, "requires start");
+      start = moment(start).format("YYYY-MM-DD hh:mm:ss");
+      end = moment(end).format("YYYY-MM-DD hh:mm:ss");
+      // require an array of values
+      selection = lodash.castArray(selection);
+      return `
+        SELECT ${selection.join(", ")}
+        FROM
+          bigquery-public-data.crypto_ethereum.logs
+        WHERE
+          block_timestamp > TIMESTAMP('${start}')
+          AND block_timestamp < TIMESTAMP('${end}')
+          AND LOWER(address)=LOWER('${contract}')
+        ORDER BY block_timestamp ASC;
+      `;
+    },
+    // transactions by contract
+    transactionsByContract(contract, start, end = Date.now(), selection = ["*"]) {
+      assert(contract, "requires contract");
+      assert(start, "requires start");
+      start = moment(start).format("YYYY-MM-DD hh:mm:ss");
+      end = moment(end).format("YYYY-MM-DD hh:mm:ss");
+      selection = lodash.castArray(selection);
+
+      return `
+        SELECT ${selection.join(", ")}
+        FROM
+          bigquery-public-data.crypto_ethereum.transactions
+        WHERE
+          block_timestamp > TIMESTAMP('${start}')
+          AND block_timestamp < TIMESTAMP('${end}')
+          AND LOWER(to_address)=LOWER('${contract}')
+        ORDER BY block_timestamp ASC;
+      `;
```

Comment:
Yea this is a really good point. The reason is that we wanted to maximize flexibility by moving all the processing in JS. You can imagine getting a different source for data, such as web3 or graphql, where then we would need a different pipeline since you cant query them this way. The way its set up now the JS is pretty source agnostic. This can be optimized like you are suggesting if we feel fine with locking in both the vendor and the tagging methodology.